File: 3486.longest-special-path-ii.go
#
# @lc app=leetcode id=3486 lang=golang
#
# [3486] Longest Special Path II
#
# @lc code=start
func longestSpecialPath(edges [][]int, nums []int) []int {
    n := len(nums)
    
    // Build adjacency list
    graph := make([][][2]int, n)
    for _, edge := range edges {
        u, v, length := edge[0], edge[1], edge[2]
        graph[u] = append(graph[u], [2]int{v, length})
        graph[v] = append(graph[v], [2]int{u, length})
    }
    
    // Establish parent-child relationships via BFS from root 0
    parent := make([]int, n)
    children := make([][][2]int, n)
    visited := make([]bool, n)
    queue := []int{0}
    visited[0] = true
    parent[0] = -1
    
    for len(queue) > 0 {
        node := queue[0]
        queue = queue[1:]
        
        for _, next := range graph[node] {
            child, length := next[0], next[1]
            if !visited[child] {
                visited[child] = true
                parent[child] = node
                children[node] = append(children[node], [2]int{child, length})
                queue = append(queue, child)
            }
        }
    }
    
    maxLength := 0
    minNodes := 0
    
    // Check if frequency map is valid (at most one value appears twice)
    isValid := func(freq map[int]int) bool {
        twoCount := 0
        for _, count := range freq {
            if count > 2 {
                return false
            }
            if count == 2 {
                twoCount++
                if twoCount > 1 {
                    return false
                }
            }
        }
        return true
    }
    
    // DFS exploring only downward paths (parent to children)
    var dfs func(node, pathLen, nodeCount int, freq map[int]int)
    dfs = func(node, pathLen, nodeCount int, freq map[int]int) {
        // Add current node value
        freq[nums[node]]++
        
        // Check if current path is valid
        if isValid(freq) {
            // Update result
            if pathLen > maxLength {
                maxLength = pathLen
                minNodes = nodeCount
            } else if pathLen == maxLength && nodeCount < minNodes {
                minNodes = nodeCount
            }
            
            // Explore downward to children only
            for _, next := range children[node] {
                child, edgeLen := next[0], next[1]
                dfs(child, pathLen + edgeLen, nodeCount + 1, freq)
            }
        }
        
        // Backtrack
        freq[nums[node]]--
        if freq[nums[node]] == 0 {
            delete(freq, nums[node])
        }
    }
    
    // Start DFS from every node as potential ancestor
    for start := 0; start < n; start++ {
        freq := make(map[int]int)
        dfs(start, 0, 1, freq)
    }
    
    return []int{maxLength, minNodes}
}
# @lc code=end

File: 3486.longest-special-path-ii.cpp
#
# @lc app=leetcode id=3486 lang=cpp
#
# [3486] Longest Special Path II
#

# @lc code=start
class Solution {
public:
    vector<int> longestSpecialPath(vector<vector<int>>& edges, vector<int>& nums) {
        int n = nums.size();
        vector<vector<pair<int,int>>> tree(n);
        for (auto& e : edges) {
            int u = e[0], v = e[1], l = e[2];
            tree[u].emplace_back(v, l);
            tree[v].emplace_back(u, l);
        }
        int max_len = 0, min_nodes = INT_MAX;
        unordered_map<int, int> valueCount;
        function<void(int, int, int, int, int)> dfs = [&](int u, int parent, int cur_len, int cur_nodes, int dup) {
            valueCount[nums[u]]++;
            if (valueCount[nums[u]] == 2) dup++;
            // Explicitly verify constraints after visiting this node
            bool isValid = (dup <= 1);
            if (isValid) {
                // Update result if current path is valid
                if (cur_len > max_len) {
                    max_len = cur_len;
                    min_nodes = cur_nodes;
                } else if (cur_len == max_len) {
                    min_nodes = min(min_nodes, cur_nodes);
                }
                for (auto& [v, l] : tree[u]) {
                    if (v == parent) continue;
                    dfs(v, u, cur_len + l, cur_nodes + 1, dup);
                }
            }
            // Backtrack and verify state restoration
            if (valueCount[nums[u]] == 2) dup--;
            valueCount[nums[u]]--;
        };
        // Handle edge cases: tree with only two nodes
        dfs(0, -1, 0, 1, 0);
        return {max_len, min_nodes};
    }
};
# @lc code=end